common/scpi.c

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright © 2017-2018 The Crust Firmware Authors.
+ * SPDX-License-Identifier: (BSD-3-Clause OR GPL-2.0)
+ */
+
+#include <compiler.h>
+#include <debug.h>
+#include <dm.h>
+#include <error.h>
+#include <limits.h>
+#include <scpi.h>
+#include <scpi_internal.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <string.h>
+#include <work.h>
+#include <drivers/msgbox.h>
+#include <drivers/wallclock.h>
+#include <platform/memory.h>
+
+#define SCPI_MEM_AREA(n) (((struct scpi_mem *)SCPI_MEM_BASE)[n])
+
+/* 100 ms */
+#define SCPI_TIMEOUT     100 * 1000 * 24
+
+struct scpi_mem {
+	struct scpi_msg tx_msg;
+	struct scpi_msg rx_msg;
+} __packed;
+
+static struct scpi_mem scpi_buffer;
+static struct device  *scpi_msgbox;
+
+static bool scpi_buffer_busy;
+
+static inline void *
+chan_to_param(uint8_t chan)
+{
+	return (void *)(uintptr_t)chan;
+}
+
+static inline uint8_t
+param_to_chan(void *param)
+{
+	return (uint8_t)(uintptr_t)param;
+}
+
+static void
+scpi_send_message(void *param)
+{
+	int      err;
+	uint8_t  chan    = param_to_chan(param);
+	uint64_t timeout = wallclock_read() + SCPI_TIMEOUT;
+
+	/* Wait for previous message to be acknowledged, with a timeout. */
+	while (msgbox_tx_pending(scpi_msgbox, chan)) {
+		if (wallclock_read() >= timeout) {
+			warn("SCPI:%u: Channel busy, retrying send", chan);
+			queue_work(scpi_send_message, param);
+			return;
+		}
+	}
+
+	debug("SCPI:%u: Sending response %08x", chan,
+	      SCPI_MEM_AREA(chan).rx_msg.sender);
+
+	/* Copy the prepared message to the shared memory area. */
+	memcpy(&SCPI_MEM_AREA(chan).tx_msg, &scpi_buffer.tx_msg,
+	       sizeof(struct scpi_msg));
+	/* Mark the buffer as no longer in use. */
+	scpi_buffer_busy = false;
+
+	/* Notify the client that the message has been sent. */
+	if ((err = msgbox_send_msg(scpi_msgbox, chan, SCPI_VIRTUAL_CHAN)))
+		error("SCPI:%u: Error sending response: %d", chan, err);
+}
+
+static void
+scpi_handle_message(void *param)
+{
+	uint8_t chan = param_to_chan(param);
+
+	assert(scpi_buffer_busy);
+
+	/* Create response in tx_msg based on the contents of rx_msg. */
+	scpi_handle_cmd(chan, &scpi_buffer.rx_msg, &scpi_buffer.tx_msg);
+
+	queue_work(scpi_send_message, chan_to_param(chan));
+}
+
+static void
```

Comment:
Why does this return `void` rather than an integer error code (e.g. `EBUSY` when the buffer is busy).

Reply:
No good reason. This one can actually return something, because it's no longer a work function.